protocol/signalfx/trace.go

```diff
@@ -33,14 +458,34 @@ var errInvalidJSONTraceFormat = errors.New("invalid JSON format; please see corr
 // Read the data off the wire in json format
 func (decoder *JSONTraceDecoderV1) Read(ctx context.Context, req *http.Request) error {
 	dec := json.NewDecoder(req.Body)
-	var input Dataformat
+
+	var input []*inputSpan
 	if err := dec.Decode(&input); err != nil {
 		return errInvalidJSONTraceFormat
 	}
+
 	if len(input) == 0 {
 		return nil
 	}
-	return decoder.Sink.AddSpans(ctx, input)
+
+	spans := make([]*trace.Span, 0, len(input))
+
+	// Don't let an error converting one set of spans prevent other valid spans
+	// in the same request from being rejected.
+	var conversionErrs error
+	for i := range input {
```

Comment:
This is now optimized for v1 spans, and an list of v2 spans then creates that many single arrays of v2 spans instead of just heading through. Please make the v2 case the optimal case and have a check before going through the list and allocating. Also, please write a benchmark for before you add the v1 code and show how much slower (hopefully minimal) it is with this code and fix anything glaring.

Reply:
With v2 case shortcutted the maximum amount, results are as follows:

```
$ benchcmp ~/tmp/old-bench ~/tmp/new-bench                                                                                  │
benchmark                   old ns/op     new ns/op     delta                                                                                                  │
BenchmarkTraceDecoder-8     44254         45838         +3.58%
```

Not too bad.

Reply:
Can you run that with --benchmem or set b.ReportAllocs() in your test? Also, so i'm 100% understanding, this is with code that i haven't seen yet? or the existing code i have seen?

Reply:
This is the results for the latest code in both this branch and the `trace-benchmark` branch, so it is all on github:

```
benchmark                   old ns/op     new ns/op     delta
BenchmarkTraceDecoder-8     1921702       1954816       +1.72%

benchmark                   old allocs     new allocs     delta
BenchmarkTraceDecoder-8     5439           5741           +5.55%

benchmark                   old bytes     new bytes     delta
BenchmarkTraceDecoder-8     472931        486657        +2.90%
```